### PR TITLE
Added timedelta feature and timezone conversion

### LINKED ISSUE #1: Implement timedelta for CLI
This will add a new feature to this which can be used like 

`worldtimebuddy --timezone UTC --delta +2hr`

The above command should return +2hrs in UTC timezone. Be careful this should only work if timezone flag is available.

Sample combinations

```
worldtimebuddy --timezone UTC --delta -2hr
worldtimebuddy --timezone UTC --delta -2min
worldtimebuddy --timezone UTC --delta +2min
worldtimebuddy --timezone UTC --delta -1day
```

### LINKED ISSUE #2: Timezone conversion to codes
This feature will allow the user to use IST instead of Asia/Kolkata and similar long form to short codes.

### LINKED ISSUE #3: Add Installation guide for worldtimebuddy
Installation guide is missing, please add how to install this cli tool.

### PR DESCRIPTION
**Pull request by bakayu**
closes #1 closes #2 closes #3

### CODE REVIEW

**Comment by sPredictorX1708** [comment]
Hi @bakayu, please update the testing done as well in the details.

File: worldtimebuddy/cli.py [hunk 1]
```diff
@@ -21,9 +37,27 @@ def cli(format, major, timezone, list_timezones):
...
+                elif 'min' in delta:
+                    minutes = int(delta.replace('min', ''))
+                    time += datetime.timedelta(minutes=minutes)
+                elif 'day' in delta:
+                    days = int(delta.replace('day', ''))
+                    time += datetime.timedelta(days=days)
+                else:
+                    click.echo("Error: Invalid time format. Use +2hr, -30min, -1day, etc.")
+                    return
+
```
Comment (sPredictorX1708) [inline worldtimebuddy/cli.py]: Consider this to add in utils

File: worldtimebuddy/cli.py [hunk 2]
```diff
@@ -4,12 +4,28 @@
...
+    'EST': 'US/Eastern',
+    'CST': 'US/Central',
+    'MST': 'US/Mountain',
+    'GMT': 'Europe/London',
+    'CET': 'Europe/Paris',
+    'EET': 'Europe/Athens',
+    'JST': 'Asia/Tokyo',
+    'AEST': 'Australia/Sydney',
+    'UTC': 'UTC'
+}
```
Comment (sPredictorX1708) [inline worldtimebuddy/cli.py]: Consider this adding a separate constants file.

Comment (bakayu) [reply worldtimebuddy/cli.py]: will get back to it right away, I worked on it before you added the utils.py file :)

File: worldtimebuddy/cli.py (hunk 1 above)
Comment (bakayu) [reply worldtimebuddy/cli.py]: can you kindly elaboraite on this? thank you.

Comment (sPredictorX1708) [reply worldtimebuddy/cli.py]: Consider moving the block 

```
if 'hr' in delta:
    hours = int(delta.replace('hr', ''))
    time += datetime.timedelta(hours=hours)
elif 'min' in delta:
    minutes = int(delta.replace('min', ''))
    time += datetime.timedelta(minutes=minutes)
elif 'day' in delta:
    days = int(delta.replace('day', ''))
    time += datetime.timedelta(days=days)
```
into some util function in utils.py file.

File: worldtimebuddy/cli.py (hunk 2 above)
Comment (sPredictorX1708) [reply worldtimebuddy/cli.py]: Also please add testing details for both the changes.

Comment (bakayu) [reply worldtimebuddy/cli.py]: > Also please add testing details for both the changes.

would screen shots be enough?

Comment (sPredictorX1708) [reply worldtimebuddy/cli.py]: Can you check the PR - https://github.com/dubesar/worldtimebuddy/pull/7, attach similarly with different cases.

**Comment by bakayu** [comment]
Updated as per your feedback :)

testing:

timedelta and usage of short timezone codes
``` bash
worldtimebuddy --timezone IST --delta +2hr
Asia/Kolkata                   2024-10-07 01:40:07
```
``` bash
worldtimebuddy --timezone JST --delta +5day
Asia/Tokyo                     2024-10-12 03:10:11
```
``` bash
worldtimebuddy --timezone UTC --delta -1min
UTC                            2024-10-06 18:09:13
```

image

File: worldtimebuddy/cli.py [hunk 3]
```diff
@@ -11,7 +12,8 @@
 @click.option('--major', is_flag=True, help='Show only major timezones')
 @click.option('--timezone', '-tz', help='Show time for a specific timezone')
 @click.option('--list', 'list_timezones', is_flag=True, help='List all available timezones')
-def cli(format, major, timezone, list_timezones):
+@click.option('--delta', help='Time to add (e.g., +2hr, -30min, -1day)')
```
Comment (sPredictorX1708) [inline worldtimebuddy/cli.py:15]: Please modify readme for new parameter and its usage.

File: worldtimebuddy/utils.py [hunk 4]
```diff
@@ -1,9 +1,24 @@
...
+        hours = int(delta.replace('hr', ''))
+        time += datetime.timedelta(hours=hours)
+    elif 'min' in delta:
+        minutes = int(delta.replace('min', ''))
+        time += datetime.timedelta(minutes=minutes)
+    elif 'day' in delta:
+        days = int(delta.replace('day', ''))
+        time += datetime.timedelta(days=days)
+    else:
+        raise ValueError("Invalid time format. Use +2hr, -30min, -1day, etc.")
```
Comment (sPredictorX1708) [inline worldtimebuddy/utils.py]: If returning a datetime how will the ValueError behave? Can you think of better approach here for return type? Also please attach testing for negative case as well where explicitly the else case gets executed.

File: worldtimebuddy/cli.py (hunk 3 above)
Comment (bakayu) [reply worldtimebuddy/cli.py:15]: on it

File: worldtimebuddy/utils.py (hunk 4 above)
Comment (bakayu) [reply worldtimebuddy/utils.py:23]: > If returning a datetime how will the ValueError behave? Can you think of better approach here for return type? 

should I define a custom error in case an invalid argument is passed and else case is executed?

Comment (sPredictorX1708) [reply worldtimebuddy/utils.py]: I think ValueError seems fine, but in that case there is no return value right? How are we handling that?

Comment (bakayu) [reply worldtimebuddy/utils.py]: > If returning a datetime how will the ValueError behave? Can you think of better approach here for return type?

when ValueError is raised its caught by the try catch block in cli.py which calls the delta function
``` py
if delta:
    try:
         time = callDelta(time, delta)
    except ValueError as e:
         click.echo(f"Error: {e}")
         return
```

I can rename it to a custom error to avoid confusion, will that do?

Comment (sPredictorX1708) [reply worldtimebuddy/utils.py]: Yes, also update the testing for that.

**Comment by bakayu** [comment]
Updated as per your feedback :)

additional testing:
```
worldtimebuddy --timezone UTC --delta +2
DeltaValueError: Invalid time format. Use +2hr, -30min, -1day, etc.
```
```
worldtimebuddy --timezone IST --delta -1day
Asia/Kolkata                   2024-10-06 00:26:31
```
```
worldtimebuddy --timezone IST --delta -1hr
Asia/Kolkata                   2024-10-06 23:26:35
```

image

**Comment by bakayu** [comment]
I have updated the README.md file to include delta function use cases and added an insallation guide as well 
this closes #3 as well now

**Comment by sPredictorX1708** [comment]
LGTM, will approve and merge after one more round of review!

File: README.md [hunk 5]
```diff
@@ -16,8 +16,22 @@ Options:
...
+### Installation:
+1. Clone the repository locally
+``` bash
+git clone https://github.com/dubesar/worldtimebuddy.git
+cd worldtimebuddy
+```
+2. Run the following code to install dependencies and worldtimebuddy
+``` bash
+pip install -r requirements.txt
+python3 setup.py built
```
Comment (sPredictorX1708) [inline README.md]: This should be `python3 setup.py build`

File: README.md [hunk 6]
```diff
@@ -75,4 +89,20 @@ Africa/Algiers
 `worldtimebuddy --timezone UTC`
 
 UTC                            2024-10-06 13:52:37
+...
+
+
+6. Getting a difference of time from a particular timezone 
+(supported time formats: +/- min, hr, day)
```
Comment (sPredictorX1708) [inline README.md]: Put a line break here.

File: README.md [hunk 7]
```diff
@@ -75,4 +89,20 @@ Africa/Algiers
...
+without --delta argument
+`worldtimebuddy --timezone UTC`
+
+UTC                            2024-10-06 18:43:35
+
+
```
Comment (sPredictorX1708) [inline README.md]: Remove one empty line from here.

File: README.md (hunk 6 above)
Comment (bakayu) [reply README.md]: the whole usage section is in a single code block, should I fix it and put individual sample code in code blocks?

File: README.md (hunk 5 above)
Comment (bakayu) [reply README.md]: my bad, fixing the typo

File: README.md (hunk 6 above)
Comment (sPredictorX1708) [reply README.md]: No just put an extra line after line#96

Comment (sPredictorX1708) [reply README.md]: > the whole usage section is in a single code block, should I fix it and put individual sample code in code blocks?

This would be great! If you can do it?

**Comment by bakayu** [comment]
updated README.md file

File: README.md [hunk 8]
```diff
@@ -16,35 +16,53 @@ Options:
...
 3. Get major timezones
 
 Currenlty only IST/PST/PDT/UTC by default. But user can set major timezones in their environment with name `MAJOR_TIMEZONES`, this tool will pick up those as the major
 versions.
+eg.
+```
+> export MAJOR_TIMEZONES=["UTC", "Asia/Kolkata"]
+> worldtimebuddy --major
 
 eg. export MAJOR_TIMEZONES=["UTC","Asia/Kolkata"]
```
Comment (sPredictorX1708) [inline README.md]: Seems duplicate

File: README.md [hunk 9]
```diff
@@ -57,22 +75,39 @@ Asia/Kolkata                   2024-10-06 19:20:26
...
+> worldtimebuddy --timezone UTC
 
 UTC                            2024-10-06 13:52:37
+...
+```
+
+6. Getting a difference of time from a particular timezone <br> (supported time formats: +/- min, hr, day) 
+`without --delta argument`
+```
+> worldtimebuddy --timezone UTC
```
Comment (sPredictorX1708) [inline README.md]: Is this example required? Can we just show an extra example in point 5? Like

```
> worldtimebuddy --timezone JST
```

??

File: README.md (hunk 8 above)
Comment (sPredictorX1708) [reply README.md]: Btw, Line#67 is correct one, Line#64 won't work.

File: README.md (hunk 9 above)
Comment (bakayu) [reply README.md]: done, please review

**Comment by bakayu** [comment]
please review.

File: README.md [hunk 10]
```diff
@@ -64,17 +64,9 @@ eg.
 > export MAJOR_TIMEZONES=["UTC", "Asia/Kolkata"]
```
Comment (sPredictorX1708) [inline README.md]: This should be MAJOR_TIMEZONES=["UTC","Asia/Kolkata"], Please make this small change, rest LGTM. Thanks

Comment (bakayu) [reply README.md]: done

File: README.md [hunk 11]
```diff
@@ -61,7 +61,7 @@ Currenlty only IST/PST/PDT/UTC by default. But user can set major timezones in t
 versions.
 eg.
 ```
-> export MAJOR_TIMEZONES=["UTC", "Asia/Kolkata"]
```
Comment (sPredictorX1708) [inline README.md]: I meant 

export MAJOR_TIMEZONES=["UTC","Asia/Kolkata"]
Please check the comment above, we cannot have space in the array.

Comment (bakayu) [reply README.md]: I see, updated with appropriate changes.

**Review by sPredictorX1708 — APPROVED** [review]
LGTM, Thanks for the change!

**Comment by sPredictorX1708** [comment]
Merged!

**Comment by bakayu** [comment]
please add the hacktoberfest-accepted label, thanks! ^^
